Current overview of Vocabularies and Contexts (#13)
w3c/vc-wg: vocabs/v2.md

## 1	Preliminary remarks

- All major vocabularies are maintained using the same tool: [yml2vocab](https://github.com/w3c/yml2vocab/): the vocabulary is defined using a special `.yml` file, and the tool generates a human-readable HTML version, as well as formal “RDFS+bit of OWL” versions in Turtle and JSON-LD serializations. In all core cases the VC specifications include a (normative) hash of the JSON-LD versions (when appropriate).

> - During When in development:

Reviewer: ```suggestion
- When in development:
```

	- the YAML versions are part of the relevant GitHub repositories;
	- the HTML, JSON-LD, and Turtle versions appear only under the "w3c.github.io/…" URL-s: these files are generated, using yml2vocab, as part of GitHub actions;
	- In some cases there is also an SVG file, which is a hand-crafted visual representation of the vocabulary displayed in the HTML file. It is on the GigHub repository.
- JSON-LD context files are maintained manually on the relevant GitHub repositories. During development the canonical URL is redirected to a "w3c.github.io/…" URL.
- At the publication of the Recommendations the HTML, Turtle, and JSON-LD versions of the vocabularies, as well as the JSON-LD Context files, are copied to the W3C server under the https://www.w3.org/2025/credentials/ date-space folder. Redirections are set up from the canonical URLs of to the final places.

> 	- Where appropriate, the structure of the directories and files on `/2025/credentials` follow the versioning. This means that the hierarchy is prepared for new versions already.

Author: ```suggestion
- At the publication of the Recommendations the HTML, Turtle, and JSON-LD versions of the vocabularies, as well as the JSON-LD Context files, are copied to the W3C server under the https://www.w3.org/2025/credentials/ date-space folder. Redirections are set up from the canonical URLs of to the final places.
	- Where appropriate, the structure of the directories and files on `/2025/credentials` follow the versioning. This means that the hierarchy is prepared for new versions already.
- In contrast to vocabularies JSON-LD context files, once finalized, should be considered as immutable. Hence a reference to their versions (e.g., `v1`) included in their URLs. Additionally, a pattern of adding a suffix of `rc1, rc2, …, rcN` for "release candidates" may have to be adopted for future developments.
```

Reviewer: I added a suggestion on top of this suggestion below.

Author: Marking this as resolved, using @dlongley 's separate suggestion

> 	- Where appropriate, the structure of the directories and files on `/2025/credentials` follow the versioning. This means that the hierarchy is prepared for new versions already.

Reviewer: +1 ... with a slight suggestion on your suggestion:

```suggestion
- At the publication of the Recommendations the HTML, Turtle, and JSON-LD versions of the vocabularies, as well as the JSON-LD Context files, are copied to the W3C server under the https://www.w3.org/2025/credentials/ date-space folder. Redirections are set up from the canonical URLs to the final places.
	- Where appropriate, the structure of the directories and files on `/2025/credentials` follow the versioning. This means that the hierarchy is prepared for new versions already.
- In contrast to vocabularies, JSON-LD context files should be considered immutable. Hence a reference to their versions (e.g., `v1`) included in their URLs. Additionally, a pattern of adding a suffix of `rc1, rc2, …, rcN` for "release candidates" during development is recommended. This naming approach allows for updates prior to finalization whilst preserving the immutability property.
```

This:

1. Adds a comma after "In contrast to vocabularies" and fixes a "of to" grammatical typo.
2. Removes the "once finalized" text because many implementers will treat *all* contexts in this space as immutable so we should do the same, even if we might need wiggle room right at the very end of CR or something because of a technical error (i.e., it's a SHOULD not a MUST).
3. More strongly "recommends" using the suffix mechanism and gives the reasoning for why.

- When in development:
	- readable via  https://w3c.github.io/vc-data-model/vocab/credentials/v2/examples.html
	- source: https://github.com/w3c/vc-data-model/blob/main/vocab/credentials/v2/examples.html
##### 2.1.1.3	"Undefined" vocabulary — https://www.w3.org/ns/credentials/undefined-terms#

- In final form it is redirected via `https://www.w3.org/ns/credentials/.htaccess` to https://www.w3.org/2025/credentials/vcdm/undefined-terms/vocab/v2/vocabulary.html (there is only an HTML file, no RDFS files)
- When in development:
	- readable via https://w3c.github.io/vc-data-model/vocab/credentials/v2/undefined-term.html
	- source: https://github.com/w3c/vc-data-model/blob/main/vocab/credentials/v2/undefined-term.html
#### 2.1.2	Context files

##### 2.1.2.1	Core context file — https://www.w3.org/ns/credentials/v2

- In final form it is redirected via `https://www.w3.org/ns/credentials/.htaccess` to https://www.w3.org/2025/credentials/vcdm/context/v2.jsonld
- When in development:
	- reachable via https://w3c.github.io/vc-data-model/contexts/credentials/v2 

> 	- source: https://github.com/w3c/vc-data-model/blob/main/contexts/credentials/v2

Reviewer: This might be mentioned for historical purposes for already-completed contexts, but once contexts are "done" they should be considered immutable, never to be changed again. We probably want to reflect that here for all contexts.

Additionally, a pattern of adding a suffix of `rc1`, `rc2`, ... `rcN` for "release candidate <number>" has been adopted in other groups to handle contexts during development. This allows bleeding edge software to be able to refer to contexts that are in development, but to also allow each new one to become an immutable context, with the final one at CR/REC time dropping the `rcN` suffix.

Author: @dlongley I think this is a remark for the introductory section, rather than here.

Author: See https://github.com/w3c/vc-wg/pull/13#pullrequestreview-4257275875

##### 2.1.2.2	"Example" context file – https://www.w3.org/ns/credentials/examples/v2

- Redirected via `https://www.w3.org/ns/credentials/.htaccess` to https://www.w3.org/2025/credentials/vcdm/examples/context/v2.jsonld

> - Development

Reviewer: ```suggestion
- When in development:
```

### 2.3	Status Lists

#### 2.3.1	Vocabulary — https://www.w3.org/ns/credentials/status#

- URL: https://www.w3.org/ns/credentials/status#
- Redirected, via `https://www.w3.org/2018/credentials/status/.htaccess` to https://www.w3.org/2025/credentials/sl/vocab/v1/vocabulary.html (and `.ttl`,`.jsonld`, and `.svg` via content negotiation)
- When in development:
	- readable via https://w3c.github.io/vc-bitstring-status-list/vocab/vocabulary.html (and `.ttl`,`.jsonld`, and `.svg`)
	- source: https://github.com/w3c/vc-bitstring-status-list/blob/main/vocab/vocabulary.yml
#### 2.3.2	Context file — https://www.w3.org/ns/credentials/status/v1

- Redirected, via `https://www.w3.org/ns/credentials/status/.htaccess` to https://www.w3.org/2025/credentials/sl/context/v1.jsonld
- When in development:
	- reachable via https://w3c.github.io/vc-bitstring-status-list/contexts/v1.jsonld
	- source: https://github.com/w3c/vc-bitstring-status-list/blob/main/contexts/v1.jsonld
### 2.4	Controlled Identity

#### 2.4.1	Vocabulary

N/a: all terms are part of the security vocabulary.

#### 2.4.2	Context file — https://www.w3.org/ns/cid/v1

> - Redirected, via `https://www.w3.org/ns/cid/.htaccess` to https://www.w3.org/2025/credentials/cid/context/v1.jsonl

Reviewer: Probably a typo:

```suggestion
- Redirected, via `https://www.w3.org/ns/cid/.htaccess` to https://www.w3.org/2025/credentials/cid/context/v1.jsonld
```

## 1	Preliminary remarks

> - All major vocabularies are maintained using the same tool: [yml2vocab](https://github.com/w3c/yml2vocab/): the vocabulary is defined using a special `.yml` file, and the tool generates a human-readable HTML version, as well as formal “RDFS+bit of OWL” versions in Turtle and JSON-LD serializations. In all core cases the VC specifications include a (normative) hash of the JSON-LD versions (when appropriate).

Reviewer 2: ```suggestion
- All major vocabularies are maintained using the same tool: [yml2vocab](https://github.com/w3c/yml2vocab/). The vocabulary is defined using a special `.yml` file, and the tool generates a human-readable HTML version, as well as formal “RDFS + a bit of OWL” versions in Turtle and JSON-LD serializations. In all core cases, the VC specifications include a (normative) hash of the JSON-LD versions (when appropriate).
```

- When in development:
	- the YAML versions are part of the relevant GitHub repositories;
	- the HTML, JSON-LD, and Turtle versions appear only under the "w3c.github.io/…" URL-s: these files are generated, using yml2vocab, as part of GitHub actions;

> 	- In some cases there is also an SVG file, which is a hand-crafted visual representation of the vocabulary displayed in the HTML file. It is on the GigHub repository.

Reviewer 3: ```suggestion
	- In some cases there is also an SVG file, which is a hand-crafted visual representation of the vocabulary displayed in the HTML file. It is on the GitHub repository.
```

> 	- the YAML versions are part of the relevant GitHub repositories;
> 	- the HTML, JSON-LD, and Turtle versions appear only under the "w3c.github.io/…" URL-s: these files are generated, using yml2vocab, as part of GitHub actions;
> 	- In some cases there is also an SVG file, which is a hand-crafted visual representation of the vocabulary displayed in the HTML file. It is on the GitHub repository.
> - JSON-LD context files are maintained manually on the relevant GitHub repositories. During development the canonical URL is redirected to a "w3c.github.io/…" URL.
> - At the publication of the Recommendations the HTML, Turtle, and JSON-LD versions of the vocabularies, as well as the JSON-LD Context files, are copied to the W3C server under the https://www.w3.org/2025/credentials/ date-space folder. Redirections are set up from the canonical URLs to the final places.
> 	- Where appropriate, the structure of the directories and files on `/2025/credentials` follow the versioning. This means that the hierarchy is prepared for new versions already.

Reviewer 2: ```suggestion
  - The YAML versions are part of the relevant GitHub repositories.
  - The HTML, JSON-LD, and Turtle versions appear only under the "w3c.github.io/…" URLs. These files are generated using `yml2vocab`, as part of the GitHub actions.
  - In some cases, there is also an SVG file, which is a hand-crafted visual representation of the vocabulary displayed in the HTML file. It is on the GitHub repository.
- JSON-LD context files are maintained manually on the relevant GitHub repositories. During development the canonical URL is redirected to a "w3c.github.io/…" URL.
- When the Recommendations are published, the HTML, Turtle, and JSON-LD versions of the vocabularies, as well as the JSON-LD Context files, are copied to the W3C server under the `https://www.w3.org/2025/credentials/` data-space folder. Redirections are set up from the canonical URLs to the final places.
  - Where appropriate, the structure of the directories and files on `/2025/credentials` follow the versioning. This means that the hierarchy is already prepared for new versions.
```

> - In contrast to vocabularies, JSON-LD context files should be considered immutable. Hence a reference to their versions (e.g., `v1`) included in their URLs. Additionally, a pattern of adding a suffix of `rc1, rc2, …, rcN` for "release candidates" during development is required. This naming approach allows for updates prior to finalization whilst preserving the immutability property.

Reviewer 2: ```suggestion
- In contrast to vocabularies, JSON-LD context files should be considered immutable. Hence, a reference to their versions (e.g., `v1`) is included in their URLs. Additionally, a pattern of adding a suffix of `rc1, rc2, …, rcN` for "release candidates" is required during development. This naming approach allows for updates prior to finalization whilst preserving the property of immutability.
```

> - In contrast to vocabularies, JSON-LD context files should be considered immutable. Hence a reference to their versions (e.g., `v1`) included in their URLs. Additionally, a pattern of adding a suffix of `rc1, rc2, …, rcN` for "release candidates" during development is recommended. This naming approach allows for updates prior to finalization whilst preserving the immutability property.

Reviewer 3: ```suggestion
- In contrast to vocabularies, JSON-LD context files should be considered immutable. Hence a reference to their versions (e.g., `v1`) included in their URLs. Additionally, a pattern of adding a suffix of `rc1, rc2, …, rcN` for "release candidates" during development is required. This naming approach allows for updates prior to finalization whilst preserving the immutability property.
```

Reviewer 3: I don't think there is a good reason to name a JSON-LD context that is in development with anything other than `rcN` suffix. Bad things happen almost immediately if we don't do that. :)

> (Note that the yml2vocab tool has evolved over the years, and it can now generate decent JSON-LD Context files as well. This could be used for newer vocabularies.)

Reviewer 3: We should shift over to using yml2vocab to generate all JSON-LD context files by default. If there are corner-cases, we should consider moving the functionality for the corner case into yml2vocab. This removes the excuse that "JSON-LD Contexts are too difficult to author".

Author: I seem to remember that some of our older context files did more than "just" represent a vocabulary. For example, they may contain references to terms not defined for that vocabulary. yml2vocab does not do that...

Definitely something to discuss

- When in development:
	- reachable via https://w3c.github.io/vc-data-model/contexts/credentials/v2 
	- source: https://github.com/w3c/vc-data-model/blob/main/contexts/credentials/v2
##### 2.1.2.2	"Example" context file – https://www.w3.org/ns/credentials/examples/v2

- Redirected via `https://www.w3.org/ns/credentials/.htaccess` to https://www.w3.org/2025/credentials/vcdm/examples/context/v2.jsonld
- When in development:
	-  reachable via https://w3c.github.io/vc-data-model/contexts/credentials/examples/v2 
	- source: https://github.com/w3c/vc-data-model/blob/main/contexts/credentials/examples/v2
##### 2.1.2.3	"Undefined" context file — https://www.w3.org/ns/credentials/undefined-terms/v2

- Redirected via `https://www.w3.org/ns/credentials/.htaccess` to https://www.w3.org/2025/credentials/vcdm/undefined-terms/context/v2.jsonld
- When in development:
	- reachable via https://w3c.github.io/vc-data-model/contexts/credentials/undefined-terms/v2 
	- source: https://github.com/w3c/vc-data-model/blob/main/contexts/credentials/undefined-terms/v2

### 2.2	Data Integrity

#### 2.2.1	Vocabulary — https://w3id.org/security#

> - Redirected via `w3cid.org` to https://www.w3.org/2025/credentials/vcdi/vocab/v2/vocabulary.html (and `.ttl`,`.jsonld`, and `.svg` via content negotiations).

Reviewer 3: ```suggestion
- Redirected via `w3id.org` to https://www.w3.org/2025/credentials/vcdi/vocab/v2/vocabulary.html (and `.ttl`,`.jsonld`, and `.svg` via content negotiations).
```
